File: dart/commection/lib/simpleHttpType/url.dart
import 'package:fast_immutable_collections/fast_immutable_collections.dart';
import 'package:meta/meta.dart';
import 'package:narumincho_util/narumincho_util.dart';

/// Structured, read-only URLs
/// where you don't have to worry about Trailing Slash or anything else
@immutable
class SimpleUrl {
  const SimpleUrl({
    required this.origin,
    required this.pathSegments,
    required this.query,
  });

  factory SimpleUrl.fromUri(Uri uri) {
    return SimpleUrl(
      origin: uri.origin,
      pathSegments: IList(uri.pathSegments.mapAndRemoveNull(
        (element) => NonEmptyString.fromString(element),
      )),
      query: IMap.fromEntries(uri.queryParameters.entries.mapAndRemoveNull(
        (entry) {
          final key = NonEmptyString.fromString(entry.key);
          return key == null ? null : MapEntry(key, entry.value);
        },
      )),
    );
  }

  final String origin;
  final IList<NonEmptyString> pathSegments;
  final IMap<NonEmptyString, String> query;
}

File: deno/simpleHttpType/url.ts
import { NonEmptyString, nonEmptyStringFrom } from "../nonEmpty.ts";

/**
 * Structured, read-only URLs
 * where you don't have to worry about Trailing Slash or anything else
 */
export type SimpleUrl = {
  readonly origin: string;
  readonly pathSegments: ReadonlyArray<NonEmptyString>;
  readonly query: ReadonlyMap<NonEmptyString, string>;
};

export const urlToSimpleUrl = (url: URL): SimpleUrl => {
  return {
    origin: url.origin,
    pathSegments: url.pathname.split("/").flatMap(
      (item): [NonEmptyString] | [] => {
        const segment = nonEmptyStringFrom(item);
        if (segment === undefined) {
          return [];
        }
        return [segment];
      },
    ),
    query: new Map([...url.searchParams].flatMap((entriy) => {
      const key = nonEmptyStringFrom(entriy[0]);
      if (key === undefined) {
        return [];
      }
      return [[key, entriy[1]]];
    })),
  };
};

export const simpleUrlToUrlText = (simpleUrl: SimpleUrl): string => {
  const url = new URL(simpleUrl.origin);
  url.pathname = "/" + simpleUrl.pathSegments.join("/");
  for (const [queryKey, queryValue] of simpleUrl.query) {
    url.searchParams.set(queryKey, queryValue);
  }
  return url.toString();
};
